### PR TITLE
feature: added public C API for 3rd-party NGINX C modules to register their own shm-based data structures for the Lua land usage

### PR DESCRIPTION
**Pull request by rainingmaster**
this pr is ported directly from [ngx_http_lua_module's feature](https://github.com/openresty/lua-nginx-module/commit/73bd64dbc145bc89c049075532ae841f7d7b6e02)

thanks for @doujiang24 's help  : )

### CODE REVIEW

File: src/api/ngx_stream_lua_api.h [hunk 1]
```diff
@@ -36,12 +38,26 @@ typedef struct {
...
     const char *package, lua_CFunction func);
 
 ngx_shm_zone_t *ngx_stream_lua_find_zone(u_char *name_data, size_t name_len);
 
+ngx_shm_zone_t *ngx_stream_lua_shared_memory_add(ngx_conf_t *cf, ngx_str_t *name,
+    size_t size, void *tag);
+
+
+static ngx_inline ngx_stream_session_t *
+ngx_stream_lua_get_session(lua_State *L)
```
Comment (doujiang24) [inline src/api/ngx_stream_lua_api.h]: I'm afraid this is wrong to use static here.

File: src/api/ngx_stream_lua_api.h [hunk 2]
```diff
@@ -36,12 +38,26 @@ typedef struct {
...
+{
+    ngx_stream_session_t    *s;
```
Comment (doujiang24) [inline src/api/ngx_stream_lua_api.h]: We'd better move this into `ngx_stream_lua_api.c`

Comment (rainingmaster) [inline src/api/ngx_stream_lua_api.h]: THX, I just change it.

**Comment by rainingmaster** [comment]
Kidding me...... I'd rollback to the original git, It still broken...

Sorry, maybe I should try to delete my git and push again

**Comment by rainingmaster** [comment]
Some problem occur in a test cases

**Pull request closed without merging**